emote/algorithms/amp.py

```diff
@@ -0,0 +1,182 @@
+from typing import Callable
+
+import torch
+
+from torch import Tensor, nn
+
+from emote import Callback
+from emote.callbacks.logging import LoggingMixin
+from emote.callbacks.loss import LossCallback
+
+
+def gradient_loss_function(model_output: Tensor, model_input: Tensor) -> Tensor:
+    """
+    Given inputs and outputs of an nn.Module, computes the sum of
+    squared derivatives of outputs to the inputs
+        Arguments:
+            model_output (Tensor): the output of the nn.Module
+            model_input (Tensor): the input to the nn.Module
+        Returns:
+            loss (Tensor): the sum of squared derivatives
+    """
+    # grad can be implicitly created only for scalar outputs
+    predictions = torch.split(model_output, 1, dim=0)
+    inputs_grad = torch.autograd.grad(
+        predictions, model_input, create_graph=True, retain_graph=True
+    )
+    inputs_grad = torch.cat(inputs_grad, dim=1)
+    inputs_grad_norm = torch.square(inputs_grad)
+    inputs_grad_norm = torch.sum(inputs_grad_norm, dim=1)
+    return torch.mean(inputs_grad_norm)
+
+
+class DiscriminatorLoss(LossCallback):
+    """
+    This loss is used to train a discriminator for
+    adversarial training.
+    """
+
+    def __init__(
+        self,
+        discriminator: nn.Module,
+        state_map_fn: Callable[[Tensor], Tensor],
```

Comment:
I think we should split this up into imitation_state_map_fn and policy_state_map_fn, as they might require different mappings (this is the case for my work for instance)

Reply:
Fixed :)